File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-3.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func eq(p *TreeNode, q *TreeNode) bool{
	if p == nil && q == nil{
		return true
	}
	if p == nil && q != nil || q == nil && p != nil{
		return false
	}
	if p.Val != q.Val{
		return false
	}

	left := eq(p.Left, q.Left)
	right := eq(p.Right, q.Right)

	return left && right
}

func isSubtree(root *TreeNode, subRoot *TreeNode) bool {
	if root == nil{
		return false
	}
	check := eq(root, subRoot)
	if check == true{ // This means we found a valid subroot
		return true
	}
	l := isSubtree(root.Left, subRoot)
	r := isSubtree(root.Right, subRoot)
	return l || r
}

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> ans(0,0);
        queue<pair<int,TreeNode*>> q;
        if(!root)
        {
            return {};
        }
        q.push({0,root});
        while(!q.empty())
        {
            TreeNode* temp = q.front().second;
            int level = q.front().first;
            q.pop();
            if(ans.size()==level){ans.push_back(0);}
            ans[level] = temp->val;
            if(temp->left)
            {
                q.push({level+1,temp->left});
            }
            

            if(temp->right)
            {
                q.push({level+1,temp->right});
            }
            
        }
        return ans;

        
    }
};
